packages/@lwc/engine-core/src/framework/hydration.ts

```diff
@@ -618,6 +622,24 @@ function validateAttrs(
     return nodesAreCompatible;
 }
 
+function checkNodeCompatibility(first: Classes, second: Classes): boolean {
```

Comment:
Bonus optimization unrelated to the empty set, the previous implementation would check all items, even though we can short circuit after the first incompatibility found.